File: ehc_customization/ehc_customization/customizations/attendance_request/attendance_request.py
import frappe
from ehc_customization.ehc_customization.customizations.attendance_request.override.override import(
validate_request_overlap, validate_timing, create_checkin_and_checkout, update_checkin_and_checkout,\
create_attendance_records, validate_allocated_permission
)
from hrms.hr.doctype.attendance_request.attendance_request import AttendanceRequest
from hrms.hr.utils import validate_active_employee, validate_dates

class AttendanceRequestCreation(AttendanceRequest):
    def validate(self):
        validate_active_employee(self.employee)
        validate_dates(self, self.from_date, self.to_date)
        self.validate_half_day()
        validate_request_overlap(self)
        validate_allocated_permission(self)
        validate_timing(self)

    def on_submit(self):
        create_attendance_records(self)
        create_checkin_and_checkout(self)
                
    def on_update_after_submit(self):
        validate_allocated_permission(self)
        update_checkin_and_checkout(self)

File: ehc_customization/ehc_customization/customizations/attendance_request/attendance_request.js
frappe.ui.form.on('Attendance Request', {
    refresh: function(frm) {
        var currentDate = frappe.datetime.now_date();
        var maxDate = frm.doc.from_date;
        if (currentDate > maxDate) {
            frm.set_df_property('end_time', 'read_only', 1);
            frm.disable_save();
            frm.set_read_only();
        }
    },
    onload: function(frm) {
        var currentDate = frappe.datetime.now_date();
        var maxDate = frm.doc.from_date;
        if (currentDate > maxDate) {
            frm.set_df_property('end_time', 'read_only', 1);
            frm.disable_save();
            frm.set_read_only();
        }
        if(frm.doc.employee && frm.doc.__islocal){
            frm.trigger('employee')
        }
        frappe.call({
            method: "ehc_customization.ehc_customization.doctype.violation_request.api.api.query_filter",
            args: {
                user: frappe.session.user_email
            },
            callback: function(r) {
				if(frappe.session.user != 'Administrator'){
					if (r.message) {
						var reportedEmployees = r.message;
                        frm.set_query('employee', function() {
							return {
								filters: [
									['Employee', 'name', 'in', reportedEmployees]
								]
							};
						});					}
				}
            }
        });
    },
    custom_reason(frm){
        frappe.call({
            method: 'ehc_customization.ehc_customization.customizations.attendance_request.override.override.update_available_hours',
            args: {
                doc: frm.doc
            },
            callback: function (response) {
                if (response.message) {
                    frm.set_value("custom_available_hours", response.message);
                }
            }
        });
    },
    employee(frm){
        if(frm.doc.employee && frm.doc.employee != '__employee_name'){
            frappe.call({
                method: 'ehc_customization.ehc_customization.customizations.attendance_request.permission.permission_query.manager_self',
                args: {
                    employee: frm.doc.employee
                },
                callback: function (r) {
                    if (r.message) {
                        frm.set_value('custom_reporting_manager', r.message);
                    } else {
                        frappe.throw(__('User Not Linked With Manager or Reporting Manager Not Assigned for this Employee'));
                    }
                }
            });
        }
    }
});